pkg/storage/types/dv.go

```diff
@@ -332,3 +334,68 @@ func (d *DataVolumeConditionManager) HasConditionWithStatusAndReason(dv *cdiv1.D
 	c := d.GetCondition(dv, cond)
 	return c != nil && c.Status == status && c.Reason == reason
 }
+
+func GetStorageClassFromCache(scName string, scStore cache.Store) (*storagev1.StorageClass, error) {
+	obj, exists, err := scStore.GetByKey(scName)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching the storageclass %s: %v", scName, err)
+	}
+	if !exists {
+		return nil, nil
+	}
+
+	sc, ok := obj.(*storagev1.StorageClass)
+	if !ok {
+		return nil, fmt.Errorf("error converting object to StorageClass: object is of type %T", obj)
+	}
+
+	return sc.DeepCopy(), nil
+}
+
+func GetCSIDriverFromCache(driver string, csiDriverStore cache.Store) (*storagev1.CSIDriver, error) {
+	obj, exists, err := csiDriverStore.GetByKey(driver)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching the csi driver %s: %v", driver, err)
+	}
+	if !exists {
+		return nil, nil
+	}
+
+	d, ok := obj.(*storagev1.CSIDriver)
+	if !ok {
+		return nil, fmt.Errorf("error converting object to storagev1.CSIDriver: object is of type %T", obj)
+	}
+
+	return d.DeepCopy(), nil
+}
+
+func IsStorageClassCSI(namespace, name string, dataVolumeStore, scStore, csiDriverStore cache.Store) (bool, error) {
+	dv, err := GetDataVolumeFromCache(namespace, name, dataVolumeStore)
+	if err != nil {
+		return false, err
+	}
+	if dv == nil {
+		return false, fmt.Errorf("datavolume %s/%s doesn't exist", namespace, name)
+	}
+	scName := ""
+	if dv.Spec.Storage != nil && dv.Spec.Storage.StorageClassName != nil {
+		scName = *dv.Spec.Storage.StorageClassName
+	} else if dv.Spec.PVC != nil && dv.Spec.PVC.StorageClassName != nil {
+		scName = *dv.Spec.PVC.StorageClassName
+	} else {
+		return false, fmt.Errorf("storage class for datavolume %s/%s is empty", namespace, name)
```

Comment:
This is actually fine. If storageClassName is not specified, the default storage class will be used.

I think you have a couple ways of handling this:

1. Lookup the default storage class. A little clunky, but doable. I'm sure CDI has code for this somewhere.

2. Get the PVC with same name as DV. With populators, the target PVC is always created, even when WFFC. The PVC will have storageClassName set. The only issue with this is non-csi + WFFC. But you don't want to support that anyway. Oh and you may have to wait a bit for the target PVC to be created

Reply:
Actually if you get the PVC, you don't have to do any of the storageclass/csi driver lookup. If a PVC is owned by a datavolume and has the dataSourceRef set, you should be good.

One thing to be aware of if you go this route is that you may have to wait for the PVC to be created so have to handle requeues

Reply:
Actually, the vm controller should be watching dvs/pvcs so shouldn't have to worry about requeue

Reply:
> This is actually fine. If storageClassName is not specified, the default storage class will be used.

It is the first thing I tried, but with this variable empty, it wasn't able to find the storage class name in the cache

Reply:
> Actually if you get the PVC, you don't have to do any of the storageclass/csi driver lookup. If a PVC is owned by a datavolume and has the dataSourceRef set, you should be good.

Not sure I get this. Even if I get the PVC , how do I know if it is from a CSI storageclass?

Reply:
We have an annotation on DVs signalling it's using CSI:
https://github.com/kubevirt/containerized-data-importer/blob/main/doc/cdi-populators.md#fallback-to-legacy-population

Reply:
@akalenyu but is the annotation always used? I interpret the documentation like is a non-CSI storageclass OR there is the annotation

Reply:
Ah, right, there's that other option, maybe disallow for that as well?

Reply:
Yes, I'm adding that option as well

Reply:
So maybe I'm misunderstanding but if we use the annotation can't we get rid of IsStorageClassCSI?

Reply:
> Not sure I get this. Even if I get the PVC , how do I know if it is from a CSI storageclass?

If the PVC is owned by a datavolume and the datasourceRef is set you can assume CSI

Reply:
> So maybe I'm misunderstanding but if we use the annotation can't we get rid of IsStorageClassCSI?

I don't see why this is necessary and would rather have KubeVirt not know about CDI annotations

Reply:
> Right, that's cleaner. But you could get rid of the CSIDriver check, right?

Yes

Reply:
@mhenriks maybe I'm not familiar enough with CDI, but why can I assume that is a CSI storage class if `datasourceRef` is set? Don't we use this field of any PVCs?

Reply:
To be clear, we are talking about PVCs owned by DataVolumes: CDI will not set `dataSourceRef` unless the storageClass supports populators (CSI). Support for populators comes from the external-provisioner CSI sidecar. Although it is technically possible that a non CSI storage class supports populators, it is unlikely and CDI will not count on it and use the "legacy" code path

Reply:
Ah I see, let me give a try

Reply:
Also I have tried, but the datasourceRef isn't populated in the DV

Reply:
datasourceRef is set on the PVC, not DV
and yeah it works with PVC field
(though you might want to use the storage field generally, not sure if you're coupled to that)